### DISCUSSION TITLE
Is it currently possible to mark multiple locations for synapse (expsyn)?

### DISCUSSION THREAD

**Discussion by lungsi** [arbor-sim/arbor#1588] [Q&A] [answered]
In the [Ring Network Tutorial](https://docs.arbor-sim.org/en/latest/tutorial/network_ring.html?)
```
# (2) Mark location for synapse at the midpoint of branch 1 (the first dendrite).
labels['synapse_site'] = '(location 1 0.5)'
...
# (3) Attach a single synapse.
decor.place('"synapse_site"', 'expsyn')
...
cell = arbor.cable_cell(tree, labels, decor)
```
can I do something like
```
labels['synapse_site_at_1'] = '(location 1 0.5)'
labels['synapse_site_at_2'] = '(location 2 0.5)'
```
and then
```
decor.place('"synapse_site_at_1"', 'expsyn')
decor.place('"synapse_site_at_2"', 'expsyn')
```
or maybe with a more compact syntax
```
labels['synapse_site'] = '(join (location 1 0.5) (location 2 0.5))'
decor.place('"synapse_site"', 'expsyn')
```
or even
```
label['synapse_site'] = '(join (region "soma") (region "dend") (region "axon"))'
decor.place('"synapse_site"', 'expsyn')
```

PS: Also, in the current version of arbor decoration is done by `arbor.decor()` but in the previous version (0.4.1.dev0) decorations could be performed with
```
cell = arbor.cable_cell(tree, labels)
leaky_chnl = arbor.mechanism('pas', {'e': -54.3, 'g': 0.0003})
cell.paint('"soma"', "na", rev_pot=50)
hh_na_k_chnl = arbor.mechanism("hh", {"gnabar": 0.120, "gkbar": 0.036)
cell.paint('"soma"', hh_na_k_chnl)
```
Is this still supported?

**Comment by noraabiakar** [comment, marked as answer]
The first 2 methods you described: 
```
labels['synapse_site_at_1'] = '(location 1 0.5)'
labels['synapse_site_at_2'] = '(location 2 0.5)'
decor.place('"synapse_site_at_1"', 'expsyn')
decor.place('"synapse_site_at_2"', 'expsyn')
```
and 
```
labels['synapse_site'] = '(join (location 1 0.5) (location 2 0.5))'
decor.place('"synapse_site"', 'expsyn')
```
will work.
This, however, will not: 
```
label['synapse_site'] = '(join (region "soma") (region "dend") (region "axon"))'
decor.place('"synapse_site"', 'expsyn')
```
This is because `synapse_site` will refer to a region on the cell, and synapses can only be placed on locsets, i.e. distinct points on the morphology.  

Also, the old way of painting mechanisms on the cell is no longer supported, the decor has replaced that interface. The benefit of this `decor` interface is that you can mix and match several `decor`, `morphology` and `label_dict`  to create different cells.

**Reply by lungsi** [reply]
Thanks @noraabiakar 

A follow-up question is if I have a cell with multiple dendritic branches what would be your advise for keeping track of multiple synaptic locations? In other words does arbor allow getting all the branch-id - needed in `(location <branch-id> 0.5)` for `"synaptic_site"` - having the same tag (say, 3 for (basal) dendrite)?

**Reply by noraabiakar** [reply]
So you want to get all branches that have tag 3, and select their midpoints? You would need to use a combo of our region and locset expressions (https://docs.arbor-sim.org/en/latest/concepts/labels.html#expressions). 
For this specific problem, one solution is to get a `locset` of the midpoints of all the branches, then restrict the `locset` to the `region` with tag 3. This is done using: `(restrict (on-branches 0.5) (tag 3))`.
- `(tag 3)` selects the `region` tagged `3`, this might include multiple complete branches or parts of branches. The region doesn't need to be connected.
- `(on-branches 0.5)` selects the `locset` of midpoints of all the branches of the morphology. 
- `(restrict locset region)` selects the locations in the `locset` that are also in the `region`.

**Reply by lungsi** [reply]
Thanks @noraabiakar this advise was very helpful :) I have one more related question.
Since, synapse in arbor is a point mechanism what if I want **synapse on every branch regardless of the tag**?

I presume one approach could be to do like what you suggested above for every tag in the cell morphology and loop through those labels during the decoration step. But is there a more effective approach?

I doubt `labels["synapse_all"] = "all"` will work since synapse is a point mechanism. Am I right?

What about `labels["synapse_all"] =  "(join (restrict (on-branches 0.5) (tag 1)) (restrict (on-branches 0.5) (tag 2)) (restrict (on-branches 0.5) (tag 3))"`?
I guess this will not work either since it returns the intersection and not points (mid points in this case) of each branch.

Based on Arbor's [DSL expression](https://docs.arbor-sim.org/en/latest/concepts/labels.html#expressions) this `(uniform (tag 3) 0 9 0)` seem to return multiple points. But they are multiple points on a specified tag not multiple points on every tag of the cell.

**Reply by noraabiakar** [reply]
If you want `expsyn` on all branches regardless of the tag, all you need is `(on-branches 0.5)`. This will return a `locset` of all the midpoints of all the branches of the cell. You can change 0.5 to whatever extent along the branch you want to place the synapses on.  
Your solution of `(join (restrict (on-branches 0.5) (tag 1)) (restrict (on-branches 0.5) (tag 2)) (restrict (on-branches 0.5) (tag 3))` works as well, but is more verbose and does the same thing (if you only have tags 1, 2, and 3 on the cell). In this case the `join` operation is a union operation on the `locsets`. 
`(uniform (tag 3) 0 9 0)` will select 10 uniformly distributed random locations (0 to 9 inclusive) on the region `tag(3)` and it will use 0 as the seed. If you don't care about the exact locations of synapses, and you just want some number of synapses randomly placed on some region, this would be the operation to use.  You can replace `(tag 3)` with any other region expression, so if you want the `locset` to be scattered over all the cell, not just the region with tag 3, you can use `(uniform (all) 0 9 0)`. (`(all)` is the region comprising the entire cell.)
(Note that if you use the same range (0 to 9) with the same seed (0) in another `uniform` operation you will get the same random locations).

**Reply by lungsi** [reply]
Great! Thank you so much @noraabiakar :+1:
